3b_parameters/MODIS_MCD12Q1_V6/1_download/download_modis_mcd12q1_v6.ipynb

```diff
@@ -233,74 +256,22 @@
-      "Successfully downloaded: MCD12Q1.A2001001.h01v11.006.2018142182942.hdf\n",
-      "Successfully downloaded: MCD12Q1.A2001001.h03v06.006.2018142183016.hdf\n",
-      "Successfully downloaded: MCD12Q1.A2001001.h00v10.006.2018142182916.hdf\n",
-      "Successfully downloaded: MCD12Q1.A2001001.h02v09.006.2018142182957.hdf\n",
-      "Successfully downloaded: MCD12Q1.A2001001.h04v10.006.2018142183058.hdf\n"
-     ]
-    }
-   ],
+   "outputs": [],
    "source": [
-    "# Loop over the download files\n",
-    "for file_url in file_list:\n",
-    "    \n",
-    "    # Make the file name\n",
-    "    file_name = file_url.split('/')[-1].strip() # Get the last part of the url, strip whitespace and characters\n",
-    "    \n",
-    "    # Check if file already exists (i.e. interupted earlier download) and move to next file if so\n",
-    "    if (modis_path / file_name).is_file():\n",
-    "        continue \n",
-    "        \n",
-    "    # Make sure the connection is re-tried if it fails\n",
-    "    retries_cur = 1\n",
-    "    while retries_cur <= retries_max:\n",
-    "        try:\n",
-    "            # Send a HTTP request to the server and save the HTTP response in a response object called resp\n",
-    "            # 'stream = True' ensures that only response headers are downloaded initially (and not all file contents too, which are 2GB+)\n",
-    "            with requests.get(file_url.strip(), verify=True, stream=True, auth=(usr,pwd)) as response:\n",
-    "        \n",
-    "                # Decode the response\n",
-    "                response.raw.decode_content = True\n",
-    "                content = response.raw        \n",
-    "        \n",
-    "                # Write to file\n",
-    "                with open(modis_path / file_name, 'wb') as data:\n",
-    "                    shutil.copyfileobj(content, data)\n",
-    "            \n",
-    "                # Progress\n",
-    "                print('Successfully downloaded: {}'.format(file_name))\n",
-    "                time.sleep(3) # sleep for a bit so we don't overwhelm the server\n",
-    "                \n",
-    "        except:\n",
-    "            print('Error downloading ' + file_name + ' on try ' + str(retries_cur))\n",
-    "            retries_cur += 1\n",
-    "            continue\n",
-    "        else:\n",
-    "            break"
+    "\n",
+    "# Download needed files with threading\n",
+    "with ThreadPoolExecutor() as executor:\n",
+    "    futures = []\n",
+    "    for file_url_raw in file_list:\n",
+    "\n",
+    "        file_url = file_url_raw.strip()\n",
+    "\n",
+    "        file_name = file_url.split('/')[-1].strip()  # Get the last part of the url, strip whitespace and characters\n",
+    "        output_file = os.path.join(modis_path, file_name)\n",
+    "        print(f'Downloading file: {file_name} from: {file_url}')\n",
+    "        futures.append(executor.submit(request_get, file_url,output_file, usr, pwd))"
```

Comment:
For the files you did manage to download with the original code, does the new code give the same results?

Reply:
It did appear so yes, though I did not do a file by file comparison. The files I downloaded did carry through successfully through the rest of the workflows.